src/crc32c.c

```diff
@@ -0,0 +1,612 @@
+ *
+ * Code retrieved in August 2016 from August 2013 post by Mark Adler on
+ *    http://stackoverflow.com/questions/17645167/implementing-sse-4-2s-crc32c-in-software
+ * Modified for use in libjulia:
+ *    - exported function renamed to jl_crc32c, DLL exports added.
+ *    - removed main() function
+ *    - architecture and compiler detection
+ *    - precompute crc32c tables and store in a generated .c file
+ *    - ARMv8 support
+ */
+
+/*
+  This software is provided 'as-is', without any express or implied
+  warranty.  In no event will the author be held liable for any damages
+  arising from the use of this software.
+
+  Permission is granted to anyone to use this software for any purpose,
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+
+  Mark Adler
+  madler@alumni.caltech.edu
+*/
+
+/* This computes a CRC-32C, *not* the CRC-32 used by Ethernet and zip, gzip, etc.
+ * A software version is provided as a fall-back, as well as for speed comparisons. */
+
+/* Version history:
+   1.0  10 Feb 2013  First version
+   1.1   1 Aug 2013  Correct comments on why three crc instructions in parallel
+*/
+
+#include "julia.h"
+#include "julia_internal.h"
+
+#ifdef _CPU_AARCH64_
+#  include <sys/auxv.h>
+#endif
+
+/* CRC-32C (iSCSI) polynomial in reversed bit order. */
+#define POLY 0x82f63b78
+
+/* Block sizes for three-way parallel crc computation.  LONG and SHORT must
+   both be powers of two.  The associated string constants must be set
+   accordingly, for use in constructing the assembler instructions. */
+#define LONG 8192
+#define LONGx1 "8192"
+#define LONGx2 "16384"
+#define SHORT 256
+#define SHORTx1 "256"
+#define SHORTx2 "512"
+
+#ifndef GEN_CRC32C_TABLES
+#include "crc32c-tables.c"
+
+#if JL_USE_IFUNC
+// Archs that can't use ifunc to do feature detection (e.g. ARM) should undef this below.
+#  define JL_CRC32C_USE_IFUNC
+#endif
+
+JL_DLLEXPORT uint32_t jl_crc32c_sw(uint32_t crci, const char *buf, size_t len);
+typedef uint32_t (*crc32c_func_t)(uint32_t crc, const char *buf, size_t len);
+
+/* Apply the zeros operator table to crc. */
+JL_UNUSED static inline uint32_t crc32c_shift(const uint32_t zeros[][256], uint32_t crc)
+{
+    return zeros[0][crc & 0xff] ^ zeros[1][(crc >> 8) & 0xff] ^
+        zeros[2][(crc >> 16) & 0xff] ^ zeros[3][crc >> 24];
+}
+
+#if (defined(_CPU_X86_64_) || defined(_CPU_X86_)) && !defined(_COMPILER_MICROSOFT_)
+#ifdef _CPU_X86_64_
+#define CRC32_PTR "crc32q"
```

Comment:
The `PTR` is the important part since it is the instruction that operates on `uintptr_t`.